trestle/core/remote/cache.py

```diff
@@ -0,0 +1,487 @@
+                raise TrestleError(
+                    f'Cache request for invalid input URI: basic authentication requires https {self._uri}'
+                )
+        self._furl.username = None
+        self._furl.password = None
+        localhost_cached_dir = self._trestle_cache_path / self._furl.host
+        # Skip any number of back- or forward slashes preceding the url path
+        fpath = str(self._furl.path)
+        path_parent = pathlib.Path(fpath[re.search('[^/\\\\]', fpath).span()[0]:]).parent
+        localhost_cached_dir = localhost_cached_dir / path_parent
+        localhost_cached_dir.mkdir(parents=True, exist_ok=True)
+        self._inst_cache_path = localhost_cached_dir / pathlib.Path(pathlib.Path(fpath).name)
+
+    def _sync_cache(self) -> None:
+        auth = None
+        if self._username is not None and self._password is not None:
+            auth = HTTPBasicAuth(self._username, self._password)
+        request = requests.get(self._furl.url, auth=auth)
+        if request.status_code == 200:
+            result = request.json()
+            if result is None:
+                raise FileNotFoundError(errno.ENOENT, os.strerror(errno.ENOENT), str(self._inst_cache_path))
+            if result['isBinary']:
+                raise NotImplementedError('Binary files are not supported!')
+            else:
+                self._inst_cache_path.write_text(result['text'])
+        else:
+            raise TrestleError(f'Query failed to run by returning code of ' f'{request.status_code}. {self._query}')
+
+
+class SFTPFetcher(FetcherBase):
+    """Fetcher for SFTP content."""
+
+    # STFP method: https://stackoverflow.com/questions/7563496/open-a-remote-file-using-paramiko-in-python-slow#7563551
+    # For SFTP fetch into memory.
+    def __init__(
+        self,
+        trestle_root: pathlib.Path,
+        uri: str,
+        settings: Settings,
+        refresh: bool = False,
+        fail_hard: bool = False,
+        cache_only: bool = False
+    ) -> None:
+        """Initialize STFP fetcher."""
+        super().__init__(trestle_root, uri, settings, refresh, fail_hard, cache_only)
+        # Is this a valid uri, however? Username and password are optional, of course.
+        u = parse.urlparse(self._uri)
+        if not u.hostname:
+            logger.error(f'Malformed URI, cannot parse hostname in URL {self._uri}')
+            raise TrestleError(f'Cache request for invalid input URI: missing hostname {self._uri}')
+        if not u.path:
+            logger.error(f'Malformed URI, cannot parse path in URL {self._uri}')
+            raise TrestleError(f'Cache request for invalid input URI: missing file path {self._uri}')
+        if u.password and not u.username:
+            logger.error(f'Malformed URI, password found but username missing in URL {self._uri}')
+            raise TrestleError(f'Cache request for invalid input URI: password found but username missing {self._uri}')
```

Comment:
Technically this is allowed as you can use the system user name.

Reply:
Ah, fair enough. I can fix that.